### PR TITLE
Add Flipper for feature flags

### PR DESCRIPTION
**Pull request by jdrosales17**
**Description:**
This PR integrates the [Flipper gem](https://github.com/jnunemaker/flipper) with the `ActiveSupportCacheStore` adapter working on top of the `ActiveRecord` adapter, so that we don't have to go to the DB to know the value of the feature flags every time.

**QA Instructions:**
To see that this is working I added a `test` feature flag that displays a text in the root path if it's enabled.

1. Go to the admin panel and log in
2. Click on the `Feature Flags` link in the sidebar menu
3. Fully enable the `test` feature flag
4. Log out of the admin panel
5. Go to the root path and check that the text is now being displayed
6. You can go back to the admin to disable the feature flag and see that the text isn't shown anymore

**Notes:**
The code added to test that this is working will be removed once it is approved to be merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -77,9 +78,10 @@ With `rails code_analysis` you can run the code analysis tool, you can omit rule
 - [Bullet](https://github.com/flyerhzm/bullet#whitelist) You can add exceptions to a bullet initializer or in the controller
 
 ## Configuring Code Climate
+
 1. After adding the project to CC, go to `Repo Settings`
-1. On the `Test Coverage` tab, copy the `Test Reporter ID`
-1. Set the current value of `CC_TEST_REPORTER_ID` in the CI project env variables
+2. On the `Test Coverage` tab, copy the `Test Reporter ID`
+3. Set the current value of `CC_TEST_REPORTER_ID` in the CI project env variables
```
Comment (martinjaimem) [inline README.md:84]:
```suggestion
1. On the `Test Coverage` tab, copy the `Test Reporter ID`
1. Set the current value of `CC_TEST_REPORTER_ID` in the CI project env variables
```

it does the trick 😉

File: app/views/dashboard/index.html.erb [hunk 2]
```diff
@@ -1,3 +1,7 @@
+<% if Flipper.enabled?(:test) %>
+  <p>Flipper is working!</p>
+<% end %>
```
Comment (jdrosales17) [inline app/views/dashboard/index.html.erb:3]: This will be removed once the PR is approved to be merged

File: config/initializers/flipper.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      adapter,
+      Rails.cache,
+      expires_in: 1.hour
+    )
+    Flipper.new(cached_adapter)
+  end
+end
+
+# The list of feature flags the application starts with
+feature_flags = ['test']
```
Comment (jdrosales17) [inline config/initializers/flipper.rb:20]: Same here, the `test` feature flag will be removed once it is approved

File: config/initializers/flipper.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+def initialize_feature_flags(feature_flags)
+  feature_flags.each do |feature_flag|
+    next if Flipper.exist?(feature_flag)
+
+    Flipper.disable(feature_flag)
+  end
+end
+
+initialize_feature_flags(feature_flags)
```
Comment (martinjaimem) [inline config/initializers/flipper.rb:30]:
```suggestion
['test'].each do |feature_flag|
    next if Flipper.exist?(feature_flag)

    Flipper.disable(feature_flag)
end
```

I'd personally remove the function as it is only being called after it is defined, but certainly not a must 👍

**Review by martinjaimem — APPROVED** [review]

File: Gemfile [hunk 5]
```diff
@@ -7,14 +7,22 @@ gem 'rails', '~> 6.1.0'
 
 # The administration framework for Ruby on Rails.
 gem 'activeadmin', '~> 2.9'
+# Use Active Model has_secure_password
+# gem 'bcrypt', '~> 3.1.7'
```
Comment (santib) [inline Gemfile:11]: why was this added?

File: Gemfile [hunk 6]
```diff
@@ -7,14 +7,22 @@ gem 'rails', '~> 6.1.0'
...
 # Adds an object-oriented layer of presentation logic to your Rails apps.
 gem 'draper', '~> 4.0', '>= 4.0.1'
+# For feature flags
+gem 'flipper-active_record', '~> 0.20.3'
+gem 'flipper-active_support_cache_store', '~> 0.20.3'
+gem 'flipper-ui', '~> 0.20.3'
 # Use Hotwire to make the app feel like an SPA
 gem 'hotwire-rails', '~> 0.1.0'
+# Use Active Storage variant
+# gem 'image_processing', '~> 1.2'
```
Comment (santib) [inline Gemfile:25]: and this

File: config/initializers/flipper.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      expires_in: 1.hour
+    )
+    Flipper.new(cached_adapter)
+  end
+end
+
+# The list of feature flags the application starts with
+feature_flags = ['test']
+
+def initialize_feature_flags(feature_flags)
```
Comment (santib) [inline config/initializers/flipper.rb:22]: I don't think we need this method

**Review by santib — APPROVED** [review]
Nice! 👏

File: db/migrate/20210202214540_create_flipper_tables.rb [hunk 8]
```diff
@@ -0,0 +1,22 @@
+class CreateFlipperTables < ActiveRecord::Migration[6.1]
+  def self.up
```
Comment (JulianPasquale) [inline db/migrate/20210202214540_create_flipper_tables.rb:2]: Can we use a `change` method instead?

Comment (jdrosales17) [reply db/migrate/20210202214540_create_flipper_tables.rb:2]: I think it is better this way, allowing to rollback.

File: config/initializers/flipper.rb (hunk 7 above)
Comment (jdrosales17) [reply config/initializers/flipper.rb:22]: Makes sense 👍

File: config/initializers/flipper.rb (hunk 4 above)
Comment (jdrosales17) [reply config/initializers/flipper.rb:30]: Makes sense, but I will keep the `feature_flags` variable so it is easier to understand.

File: db/migrate/20210202214540_create_flipper_tables.rb (hunk 8 above)
Comment (JulianPasquale) [reply db/migrate/20210202214540_create_flipper_tables.rb:2]: Using change your migration is still reversible, but I think you have a point 😄

**Review by JulianPasquale — APPROVED** [review]

File: db/migrate/20210202214540_create_flipper_tables.rb (hunk 8 above)
Comment (jdrosales17) [reply db/migrate/20210202214540_create_flipper_tables.rb:2]: Yeah, I know what you mean... Maybe this case is reversible cause it is adding 2 tables, but I don't see harm in having the `up` and `down` methods anyway.

File: Gemfile [hunk 9]
```diff
@@ -7,14 +7,22 @@ gem 'rails', '~> 6.1.0'
 
 # The administration framework for Ruby on Rails.
 gem 'activeadmin', '~> 2.9'
+# Use Active Model has_secure_password
```
Comment (brunvez) [inline Gemfile:10]: Why are these commented gems added?

File: db/migrate/20210202214540_create_flipper_tables.rb (hunk 8 above)
Comment (brunvez) [reply db/migrate/20210202214540_create_flipper_tables.rb:2]: Does `down` work correctly? We are also creating indexes but I guess `drop_table` will cascade those

**Review by juan-apa — COMMENTED** [review]
I think there is a config error, because when this is being deployed, it's failing, because the flipper gem cannot find a database in localhost. Is there somewhere where you can use the database used by the rails app?

File: db/migrate/20210202214540_create_flipper_tables.rb (hunk 8 above)
Comment (jdrosales17) [reply db/migrate/20210202214540_create_flipper_tables.rb:2]: Yep, it's working, `drop_table` handles everything.

File: Gemfile (hunk 9 above)
Comment (jdrosales17) [reply Gemfile:10]: They were already there, I just reorder them alphabetically :)
